pkg/resource/firewall/hook.go

```diff
@@ -26,4 +35,259 @@ var (
+	ko := desired.ko.DeepCopy()
+
+	rm.setStatusDefaults(ko)
+
+	if delta.DifferentAt("Spec.LoggingConfiguration") {
+		if err := rm.syncLoggingConfiguration(ctx, desired, latest); err != nil {
+			return nil, fmt.Errorf("%v: %v", err, ErrSyncingPutProperty)
+		}
+	}
+
+	return &resource{ko}, nil
+}
+
+// addPutFieldsToSpec will describe logging config and add its
+// returned values to the Firewall spec.
+func (rm *resourceManager) addLoggingConfigToSpec(
+	ctx context.Context,
+	r *resource,
+	ko *svcapitypes.Firewall,
+) (err error) {
+	getLoggingConfigurationResponse, err := rm.sdkapi.DescribeLoggingConfigurationWithContext(ctx, rm.getLoggingConfigurationPayload(r))
+	if err != nil {
+		return err
+	}
+	ko.Spec.LoggingConfiguration = rm.setResourceLoggingConfiguration(r, getLoggingConfigurationResponse)
+	return nil
+}
+
+func customPreCompare(
+	a *resource,
+	b *resource,
+) {
+	if a.ko.Spec.LoggingConfiguration == nil && b.ko.Spec.LoggingConfiguration == nil {
+		return
+	}
+
+	if a.ko.Spec.LoggingConfiguration == nil {
+		a.ko.Spec.LoggingConfiguration = &svcapitypes.LoggingConfiguration{}
+	}
+
+	if b.ko.Spec.LoggingConfiguration == nil {
+		b.ko.Spec.LoggingConfiguration = &svcapitypes.LoggingConfiguration{}
+	}
+
+	sort.Slice(a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[:], func(i, j int) bool {
+		return *a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[i].LogType < *a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[j].LogType
+	})
+	sort.Slice(b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[:], func(i, j int) bool {
+		return *b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[i].LogType < *b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[j].LogType
+	})
+}
+
+func (rm *resourceManager) getLoggingConfigurationPayload(
+	r *resource,
+) *svcsdk.DescribeLoggingConfigurationInput {
+	res := &svcsdk.DescribeLoggingConfigurationInput{}
+	res.SetFirewallName(*r.ko.Spec.FirewallName)
+	return res
+}
+
+func (rm *resourceManager) newLoggingConfigurationPayload(
+	r *resource,
+) *svcsdk.UpdateLoggingConfigurationInput {
+	res := &svcsdk.UpdateLoggingConfigurationInput{}
+	res.SetFirewallName(*r.ko.Spec.FirewallName)
+	if r.ko.Spec.LoggingConfiguration != nil {
+		res.SetLoggingConfiguration(rm.newLoggingConfiguration(r))
+	} else {
+		res.SetLoggingConfiguration(&svcsdk.LoggingConfiguration{})
+	}
+	return res
+}
+
+func makeLogDestinationConfigMapKey(cfg *svcapitypes.LogDestinationConfig) string {
+	keys := make([]string, 0, len(cfg.LogDestination))
+	for k := range cfg.LogDestination {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	key := "LogDestination:"
+	for _, k := range keys {
+		v := *cfg.LogDestination[k]
+		key += fmt.Sprintf("%s:%s,", k, v)
+	}
+
+	if cfg.LogDestinationType != nil {
+		key += fmt.Sprintf("LogDestinationType:%s,", *cfg.LogDestinationType)
+	}
+	if cfg.LogType != nil {
+		key += fmt.Sprintf("LogType:%s", *cfg.LogType)
+	}
+	return key
```

Comment:
My two cents: to avoid building strings manually with hard coded stuff, we can use `json.Marshal` to transform the map in into a "sorted" map - checkout https://go.dev/src/encoding/json/encode.go#L344

Reply:
We dont really need to 'sort' the map here. So, I am not sure how I can make use of marshaling here. I explained the logic of `compareLoggingDestinationConfigs` and associated functions below. Please let me know if you have suggestions.

Reply:
But i believe that you're calling `sort.Strings(keys)` to have deterministic output.. because of the hash map random access pattern... right? the idea was instead of manually building your "string" that is just a way of  representating of the object, how about:
```
b1, err := json.Marshall(cfg1)
b2, err := json.Marshall(cfg2)
```

then you case use string(b1), string(b2) as a key for the map discussed below

Reply:
Thank you for the suggestion. I made a change to use `json.Marshall`